Add @@ syntax for inline splatting as hashtable as new experimental feature PSGeneralizedSplatting

bergmeister opened the pull request:
# PR Summary

This is the first feature of [RFC002](https://github.com/PowerShell/PowerShell-RFC/blob/master/2-Draft-Accepted/RFC0002-Generalized-Splatting.md) for enhanced splatting.
As agreed in the accepted RFC, one should not need to declare a variable in order to splat. This PR adds the ability to definite the splatted hashtable in the same line as the command. As discussed, the `@@` has to be introduced to differentiate from the scenario where the 1st argument is a hashtable.

The scenario that this PR enables is:
```pwsh
Get-ChildItem @@{Path='C:\'; Filter=$filter}
```

This PR adds the new `At`  `TokenKind`  in order to be able the parse the syntax and adds a public `Splatted` property to `HashTableAst` so that the compiler knows at runtime that the argument is splatted.

The experimental feature `PSGeneralizedSplatting` is created for it. Although there are switch statements using the new `At` that are not guarded by an if check for this experimental feature (which was not possible due to the flow analysis of C# resulting in a compiler error), this is OK because the experimental feature flag prevents the tokenizer from creating an `At` token in the first place.

A semantic check was added so that the only implemented use case is using `@@{` as an argument to a command, therefore the user will get an actionable error when trying to use it in another context, e.g. for a .Net method or a hashtable as `@@{'key='value'}` to conform with previous behaviour (where the tokenizer returned an error in those cases). I've also checked it does not impact using `@@` inside a function name (even before a function name could not start with `@@` so no change here).

This PR does not add tab completion functionality to the keys of the splatted hashtable. This is something that is not present at the moment anyway with splatted variables and is a separate enhancement that one could add in another PR, as well as other features in the RFC. The goal of this RFC is to implement just 1 feature in this RFC that adds value, therefore making it a minimum viable feature.

I added tests for existing and new splatting functionality and error cases.

## PR Context

https://github.com/PowerShell/PowerShell-RFC/blob/master/2-Draft-Accepted/RFC0002-Generalized-Splatting.md

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name: `PSGeneralizedSplatting`
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

bergmeister committed:
Add AtAtCurly token and allow parsing of it. TODO: return command elements instead of just hashtable

bergmeister committed:
Add Splatted property to hashtable to be able to tell compiler that hashtable is an argument.

bergmeister committed:
Add tests for existing and new splatting functionality.

bergmeister committed:
restart-ci due to sporadic mac failure

bergmeister committed:
Fix the 2 CodeFactor warning (XML comments of properties must start with 'Gets or sets a value indicating whether' and a missing newline after a closing brace).

bergmeister committed:
Add CI tag to Pester tests to fix build and fix indentation

TylerLeonhardt commented:
This should probably be marked as an experimental feature.

bergmeister commented:
- I fear that if it is an experimental feature, it will not get much test coverage or usage. Usage will be limited to people can write scripts that run on v7 anyway.
- Editor syntax highlighting and code analysis in VS-Code will be cumbersome because one would need to enable the experimental feature just for viewing the code (which might be the code from another repo). Figuring out the cause of this, will not be clear to most people (the parser gives a `UnrecognizedToken` error : `Unrecognized token in source text.`).
- We still have a few previews to go until PS 7 goes RTM and the code change is a very subtle addition without explicitly altering current behaviour, therefore I see it as low risk anyway.

KirkMunro commented:
I'd also like it marked as experimental. Experimental isn't just about risk. It's about acceptance of syntax, workflow, etc., and most importantly, it lets users know that the design of this feature is not yet complete.

As expressed [here](https://github.com/PowerShell/PowerShell-RFC/issues/209) I think `@@` and `@$` are really bad additions to the PowerShell syntax, and proposed what I think are better alternatives to each of those.

KirkMunro commented:
Also, experimental does not mean it's not going to make it into v7 either. It simply means we're experimenting with some syntax/commands/etc. and want to be able to allow the broader community kick the tires on those things before we sign-off on them as supported.

SteveL-MSFT commented:
@KirkMunro is correct that Experimental means that the design can change and would not be a breaking change.

bergmeister commented:
Ok, we have 3 changes here:
- Adding to the `TokenKind`: as far I understand the `ExperimentalFeature` implementation does not allow to exclude the added enumeration item
- Usage of the added TokenKind in the Parser and Tokenizer. What should the parser do if the feature is not enabled? Should they throw explicitly if usage of the new `TokenKind` is detected or just be a no-op?
- Compiler (runtime): Usage of `ExperimentalFeature` is clear.

Happy to add the experimental feature fences around those section after the first review approves the current format.

TylerLeonhardt commented:
@daxian-dbw can probably supply advice here on proper Experimental Feature usage.

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -6933,6 +6935,14 @@ private ExpressionAst HashExpressionRule(Token atCurlyToken, bool parsingSchemaE
+            bool splatted = false;
+            if (atCurlyToken.Kind == TokenKind.AtAtCurly)
+            {
+                NextToken(); // to skip the first '@' to allow the parsing of the hashtable
```
Here where we've seen `AtAtCurly` and we now skip a token, there's a problem because `@@` might be followed by other tokens and we don't check. This causes, for example:

<img width="571" alt="Screen Shot 2019-07-09 at 21 11 36" src="https://user-images.githubusercontent.com/7009879/60939893-49e04100-a28e-11e9-8948-20ed9afa6e35.png">

My recommendation is to restructure the tokens, which I've discussed in a comment in the tokeniser.

But also there are two important things to consider:
- A `@@{` token should only be valid as an argument to a command. This should be a semantic check.
- Adding a new token, error checking and recovery is critical, since otherwise the grammar may have new pitfalls in it. The easiest solution here is to make `@@{` its own whole token.

bergmeister replied:
OK, I've changed it to perform a look-ahead so that `@{{` is correctly parsed without ambiguity.

rjmholt commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -1342,3 +1342,35 @@ foo``u{2195}abc
 }
+
+
+Describe 'Splatting' -Tags 'CI' {
```
For a language feature, we need pretty intensive testing not just of the successful invocations but of error modes (such as parse errors) as well. For splatting in particular, given the complexity, I think the bar should be set high on testing.

bergmeister replied:
I agree but I'd like to defer this to the end when the implementation becomes final. Maybe it would be great to write down a couple of test cases as requirements that could be manually tested during development?

rjmholt reviewed (changes requested):
There are some nice first steps here I think but also some fixes required.

I'm personally skeptical about implementing half a language feature (since there is always a risk that the other half never makes it in and partial grammar changes often break the full scenario from being completed), although looking at the splatting RFC it seems that it does break nicely into two things.

Another thing that should come in as part of this change is completions on the splatted hashtable itself, since completions are such a core part of the PowerShell experience.

The testing probably needs to be built out significantly to cover things like:
- If hashtables are tokenised differently, it will need to cover `@@` used in other places
- Failure semantics if the splatted hashtable expression produces a runtime error
- Errors produced if a spatted hashtable is used not as a command argument
- What happens if you try splatting a hashtable to a native command
- Likely a number of other cases I haven't thought about

lzybkr commented on src/System.Management.Automation/engine/parser/token.cs:
```diff
@@ -1119,17 +1124,19 @@ public static class TokenTraits
             /*              Command */ "command",
             /*               Hidden */ "hidden",
             /*                 Base */ "base",
+            /*            AtAtCurly */ "@@splat",
```
I was thinking of `@` as a unary operator.

Thinking that way, one could write `@ @{...}` or `@@{...}` to mean the same thing, though the later would likely be preferred.

Given that, I wouldn't expect new tokens of the form `@@(` or `@$` or `@@{`.

bergmeister replied:
I would've put it into the `Punctuators` section, similar to `AtCurly` but because it's range for the `TokenKind` enum is already full, I just decided to put it at the end. Shall I put it into the operators section then?
If I understand correctly, you'd want me to add a TokenKind for `@` instead of `@@{`and starting from `@` figure out how to parse if it is followed by`@{`?

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into InlineSplatting_AtAtCurly

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into InlineSplatting_AtAtCurly

bergmeister commented on src/System.Management.Automation/engine/parser/token.cs:
```diff
@@ -1119,17 +1124,19 @@ public static class TokenTraits
-                               "Table size out of sync with enum - _staticTokenFlags");
-            Diagnostics.Assert(s_tokenText.Length == ((int)TokenKind.Base + 1),
-                               "Table size out of sync with enum - _tokenText");
+            var maximumEnumValueOfTokenKind = Enum.GetValues(typeof(TokenKind)).Cast<int>().Max();
```
In general I would agree but because the statements are already inside a `#if DEBUG` pragma (see 3 lines above), the `static TokenTraits()` method is only executed in Debug builds anyway

bergmeister committed:
Convert if statement to switch expression to address first PR comment

bergmeister commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -3536,10 +3536,14 @@ public object VisitCommand(CommandAst commandAst)
 
-                    VariableExpressionAst variableExpression = splatTest as VariableExpressionAst;
-                    if (variableExpression != null)
+                    if (splatTest is VariableExpressionAst variableExpressionAst)
```
Ok, done now.

bergmeister committed:
Merge branch 'master' into InlineSplatting_AtAtCurly

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly_2

bergmeister committed:
Perform lookahead in tokenizer to correctly detect @@{ in all cases. The following still produces a hashtable though (it should probably throw an error, which the RTM version does atm): @@{'a'='b'}

bergmeister committed:
Make Splatted property readonly but injecting it in the constructor.

bergmeister commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -9601,6 +9601,11 @@ public override Ast Copy()
+        /// <summary>
+        /// Gets or sets a value indicating whether inline splatting syntax (@@) was used, false otherwise.
+        /// </summary>
+        public bool Splatted { get; set;  }
```
Since `HashtableAst` inherits from `ExpressionAst`, we can always change that later without a breaking change. I think we should only increase the areas where it is exposed on a need by need basis unless you can think of a case where the change in this PR could make it useful to have the property on `ExpressionAst` as well.

bergmeister commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -4613,6 +4613,11 @@ internal Token NextToken()
                         return ScanHereStringExpandable();
                     }
 
+                    if (c1 == '@')
```
I've changed it to perform a lookahead so that it correctly recognises `@{{` only at `AtAtCurly`

bergmeister commented:
@rjmholt @lzybkr I've reacted to all the comments and implemented changes: I've changed it to perform a look-ahead so that it correctly recognises `@{{` only as `AtAtCurly`.
I am aware that implementing the experimental feature and more complex tests are still outstanding but would like to get feedback or sign off on the current design. In terms of future timeline I'd like to keep this PR as minimal as possible. For example, at the moment people already use splatting despite that lack of tab completion or being able to inline it. The aim of this PR is to add only the in-lining capability to get this into PS 7, which is the most important aspect IMHO. Any hints on how to implement tab completion later would be appreciated though.

bergmeister commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -6933,6 +6935,14 @@ private ExpressionAst HashExpressionRule(Token atCurlyToken, bool parsingSchemaE
+            bool splatted = false;
+            if (atCurlyToken.Kind == TokenKind.AtAtCurly)
+            {
+                NextToken(); // to skip the first '@' to allow the parsing of the hashtable
```
I found that it did not work when using `SkipToken` as it crashes some other logic in PowerShell just when typing `@{{`

KirkMunro commented:
If possible, I'd like this design/functionality revisited by the committee. It only simplifies one scenario that would have previously required line continuation characters, and with poorly designed syntax. More effort in a design that is more respectful to how parameters are normally used while considering other line continuation scenarios would be appreciated here. This new syntax isn't sugar by any means...it's just sour.

bergmeister commented:
As far as I understand the need for the additional character in `@{{` is due to the fact that we do want to differentiate from the case of a hashtable being passed as an (unnamed) parameter, thus the design avoids introducing a breaking change. Apart from using a different character or character set I cannot think of a better way to deal with it.
I'd be happy adapting to a slightly different syntax or if you can think of a totally different approach that is better but in this case make a concrete proposal please (because this is the hardest part). With that in mind, I'd appreciate if we can make a timely decision so that a minimum viable implementation of the syntax can potentially ship in 7.0 (pending feedback). Although we've already agreed to a syntax, I'd be happy to revise but we'd need to be relatively quick, therefore some concrete proposals would help. At the end, personally I'd rather have something than nothing in 7.0. I don't find the current syntax especially bad, when using splatting with a variable one needs to do `@$` so the additional @ makes even logical sense to me.

KirkMunro commented:
> I'd be happy adapting to a slightly different syntax or if you can think of a totally different approach that is better but in this case make a concrete proposal please (because this is the hardest part).

I've already made that proposal in [PowerShell/PowerShell-RFC #179](https://github.com/PowerShell/PowerShell-RFC/pull/179), which proposes using `@` as an indicator that you want PowerShell to stop treating newlines as statement terminators until it encounters a statement terminator or a blank line. Some folks have expressed that they don't like the lack of a closing sigil with that, which is why [this alternative proposal was very recently added](https://github.com/PowerShell/PowerShell-RFC/blob/d7f3899abd789ad1fd487185ddf97195f7185806/1-Draft/RFCNNNN-Multi-Line-Continuation.md#enclosures-instead-of-a-sigil), which suggests that `` @` `` and `` `@ `` could be enclosures to enable multi-line parsing of the current statement.

The key advantage to those proposals compared with inline splatting is simple: they tell the parser how to treat newlines and leading or trailing whitespace when it comes to terminating statements, allowing users to format their code in a way that suits them better, while preserving basic PowerShell syntax. Inline splatting, on the other hand, only takes parameters into consideration, and completely changes how you would have to enter them to span a command across multiple lines without using backticks.

Commit:
Merge branch 'master' into InlineSplatting_AtAtCurly

Commit:
Add experimental feature PSGeneralizedSplatting. Due to compiler issues, some switch statements around AtAtCurly could not be guarded but that is OK since the tokeniser will never produce AtAtCurly if the experimental feature isn't enabled

Commit:
Resolve merge conflict
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

# Conflicts:
#	src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs

Commit:
undo one accidental local change in last commit

bergmeister commented:
@PoshChan  Rerun failures

PoshChan commented:
@bergmeister, all requests start with the magic word: `Please`

<details>
<summary>Commands available in this repo for you:
</summary>
<ul><li>
<code>retry &lt;target&gt;</code> this will attempt to retry only the failed jobs for the target pipeline, <code>restart</code> can be used in place of <code>retry</code>
</li>
<li>
<code>rebuild &lt;target&gt;</code> this will perform a complete rebuild of the target pipeline, <code>rerun</code> can be used in place of <code>rebuild</code> Supported values for &lt;target&gt; which can be a comma separated list are: <code>static</code>,<code>windows</code>,<code>macos</code>,<code>linux</code>,<code>all</code>
</li>
<li>
<code>get failures</code> this will attempt to get the latest failures for all of the target pipelines
</li>
<li>
<code>remind me in &lt;value&gt; &lt;units&gt;</code> this will create a reminder that will be posted after the specified duration &lt;value&gt; is a number, and &lt;units&gt; can be <code>minutes</code>, <code>hours</code>, or <code>days</code> (singular or plural)
</li>

</ul>
</details>

bergmeister commented:
@PoshChan Please retry Windows

PoshChan commented:
@bergmeister, successfully started retry of `PowerShell-CI-Windows`

bergmeister committed:
Use semantic check to restrict usage to command arguments and add more test cases. TOOD: Rename AtAtCurly to AtAt

bergmeister committed:
Rename AtAtCurly to AtAt to better represent what it is

bergmeister committed:
Cleanup using statement and whitespace

bergmeister commented:
@PoshChan  Please retry Linux

PoshChan commented on src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:
```diff
@@ -114,7 +114,10 @@ static ExperimentalFeature()
                     description: "New parameter set for ForEach-Object to run script blocks in parallel"),
```
@bergmeister, your last commit had 1 failures in `PowerShell-CI-linux`
Validate Update-Help from the Web for all PowerShell modules for user scope..Validate Update-Help for module 'Microsoft.PowerShell.Host' with scope as 'False'
```powershell
Expected 1, but got 0.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Help/UpdatableHelpSystem.Tests.ps1: line 181
181:     $helpFilesInstalled.Count | Should -Be $expectedHelpFiles.Count
```

bergmeister replied:
Sporadic failure, went away after a retry

PoshChan commented:
@bergmeister, successfully started retry of `PowerShell-CI-Linux`

bergmeister commented:
Everyone, especially @rjmholt @lzybkr  :
I've made the following changes and would consider the PR ready for merge now and would like to get feedback. I've updated the PR summary, please re-read it. What I've done after the first review:
- Add the experimental feature `PSGeneralizedSplatting`, which makes sure no `AtAt` token gets produced if the feature is not enabled and safe-guard other changed areas
- Change the `AtAtCurly` token to `AtAt` only (`@@`) and perform look-ahead in the tokenizer for a following `{` in order to mark it as a token of `AtAt` `TokenKind`.
- Add a semantic check to make sure the usage of it is restricted only to be the arguments of a `CommandAst`
- Added more positive and negative testsQ

rjmholt commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -4622,6 +4622,18 @@ internal Token NextToken()
+                    {
+                        if (c1 == '@')
+                        {
+                            var c2 = PeekChar();
```
Rather than doing a lookahead here, the tokenizer shouldn't be contextual. Instead, the token should just be `@@{`. That accurately represents the token the user sees, the lexical work done here and prevents the parser from having to skip a brace token.

It also means that the few tools that rely on PowerShell tokens don't get confused by an open brace token, rather than a single token representing `@@{`

bergmeister replied:
Ok, I've implemented your suggested change.

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

bergmeister committed:
Do not use lookahead in tokenizer so that the token is @@{

rjmholt commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -9699,6 +9701,11 @@ public override Ast Copy()
         internal bool IsSchemaElement { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether inline splatting syntax (@@) was used, false otherwise.
```
```suggestion
        /// Gets or sets a value indicating whether inline splatting syntax (using @) was used, false otherwise.
```

rjmholt reviewed (approved):
LGTM

bergmeister committed:
Update src/System.Management.Automation/engine/parser/ast.cs

Co-Authored-By: Robert Holt <rjmholt@gmail.com>

lzybkr commented:
I thought I suggested previously that there shouldn't be a `AtAt` token if this is truly generalized splatting.

Instead, `@` is like an operator that is applied to any expression, so for example, `@ @{}` (note the whitespace) generates a UnaryExpressionAst with the child HashtableAst.

KirkMunro commented:
> I thought I suggested previously that there shouldn't be a `AtAt` token if this is truly generalized splatting.
> 
> Instead, `@` is like an operator that is applied to any expression, so for example, `@ @{}` (note the whitespace) generates a UnaryExpressionAst with the child HashtableAst.

That approach could prevent other uses of `@` without risking breaking changes though, right? What are the benefits of having `@` be an unary operator for any expression vs having `@@{` as a token to open inline splatting?

I'm still hoping we'll drop this approach and move forward with `` @` `` and `` `@ `` as enclosures for multi-line continuation, but to even build that as experimental, I worry that a unary `@` operator would complicate things.

rjmholt commented:
> Instead, @ is like an operator that is applied to any expression, so for example, @ @{} (note the whitespace) generates a UnaryExpressionAst with the child HashtableAst

Yeah that makes sense to me.

@bergmeister I'm thinking that will change a few things in the PR:

- The tokenizer should look for an `At` token. It will need to look ahead to make sure it's not seeing `@(` or `@{`
- That token will need to be recognised in the current place we look for splatting, but should call expression rule underneath it I'm guessing, and generate a `SplatExpressionAst` around the expression
- That AST will mean we need new visitor methods on ICustomAstVisitor2 and AstVisitor2, with a default implementation on ICustomAstVisitor2
- Instead of trying to go through and create constructors for each expression type that makes sense as a splat, it's probably easiest to just have a `SplatExpressionAst` take an `ExpressionAst` as a child and add a semantic check to ensure there's nothing strange there. Although it might also be worth it to make it a syntactic error -- don't have a strong feeling on that one.
- Then in the compiler under `VisitCommand` you probably want something like:

```csharp
...

if (splatTest is SplatExpressionAst splatExpression)
{
    splatted = true;
}
else if (splatTest is VariableExpressionAst variableExpression)
{
    splatted = variableExpression.Splatted;
}

...
```

And `VisitSplatExpression` would probably just strip off the splat and visit the expression underneath.

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

# Conflicts:
#	src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs

bergmeister committed:
Change AtAtCurly token to At token

bergmeister commented:
I've changed the token to be just an `At` token as suggested, which I agree with is a good idea as it will keep it generic and avoid future complications of assigning certain expressions like `@@{` or `@$` to a certain token, especially since there is already a `SplattedVariable` token that will be a tad awkward to deal with in future implementations of more features of generalized splatting....
However, I disagree trying to create something where one can have a whitespace after the `At` token not only because there is no value to the user for that, but also because it unnecessarily complicates it (a lot) and was also not part of the spec per RFC. I'd like to keep the implementation to be minimum viable.
The current solution of having the `Splatted` property on the `HashTableAst` fits neatly into the existing structure and allows to lift the property to an `ExpressionAst` for future implementations of other features of generalized splatting like e.g. `Get-Item @$var` or `Get-Item @$(Get-GetItemArgument)`.
If you still disagree with me, please elaborate more why having an AST of it's own would be useful and how it would fits into existing syntax like `Get-Item @var` (which generates a `SplattedVariable` token ) without the risk of breaking things.
The current implementation has the advantage that it only adds to the code without having to touch the existing logic. I am happy to adapt the PR further but it needs more detailed reasoning. Whilst the idea of having the `At` token generate an `UnaryExpressionAst` with the HashtableAst` sounds good from a high level, I don't see how it fits into the low level design/implementation, which still has to deal with already existing syntax (which we want to touch as little as possible to avoid regressions). Or maybe I just need a hint as to where you'd image the search for the `At` token to be planted in `Parser.cs` to generate the `UnaryExpressionAst` to keep things working in the same way.

TylerLeonhardt commented:
This issue has changes that could potentially change syntax. Please consider adding this feature to [EditorSyntax](https://github.com/PowerShell/EditorSyntax) which is used for syntax highlighting in GitHub, Visual Studio Code, Atom, Sublime Text, and many more locations.

Consistent syntax highlighting is very important for the language and a feature isn't "complete" until syntax highlighting is what is expected.

If you can't contribute to [EditorSyntax](https://github.com/PowerShell/EditorSyntax), at least open an issue to track the work - however, please note, that no one is actively working on the repo and so the work will likely not get done in a timely manner.

We hope that you consider contributing to [EditorSyntax](https://github.com/PowerShell/EditorSyntax).

(note this is copy/pasted text for any change that looks like it could impact EditorSyntax - and will be a bot in the future)

KirkMunro commented:
I need to call out something about this PR that still doesn't sit right with me.

Based on my understanding this PR is designed to solve one specific issue: being able to use inline splatting of a hashtable. Does it also add support the following?

1. Inline splatting of an array?
1. Inline splatting of a hashtable that is in a parameter of an object or behind an invocation of a command?

Even if this PR supports the first scenario (via `@@(...)`) and the second scenario (via `@$(...)`), this syntax still doesn't feel like the right solution for this problem.

I suggested it before (during the original RFC if I recall correctly), but I'll bring it up again: why use this syntax instead of a common `-splat` parameter? The `-splat` parameter is respectful of PowerShell's "way", it helps advance learning/discovery (it says "splat", so that gives you something to work with when you search for help/docs), and it could even have a special-case alias of `@` if we wanted so that users could use `-@` (special case because today PowerShell does not recognize `@` as a parameter token, but we could allow it for this specific scenario if we wanted to).

That's almost the syntax that you have been pushing forward here, but supporting splatting of any array or hashtable whether it is provided as a literal value, or comes from a parameter, method, or command invocation. The only difference is you'd have a space between `-@/-splat` and the actual hashtable, which I think is also a good thing.

bergmeister commented:
@TylerLeonhardt  I've created the following tracking issue for EditorSyntax:
https://github.com/PowerShell/EditorSyntax/issues/188

bergmeister commented:
@SteveL-MSFT 
Can you please bring up this PR with  @PowerShell/powershell-committee to make a decision whether the committee wants this feature to be part of the next preview, which is the last point of being able to get features in. I have addressed comments and implemented a minimum viable version, I did reject the feature ask for being able to have a whitespace between the `@` characters and am still waiting for an answer as to why this would be essential and the RFC also does not mention such a requirement.

Currently this PR implements inline splatting of a hashtable, which is a subset of the generalized splatting RFC, however, I do not plan to implement more features as this is the minimum viable implementation of the only feature that I want. Please let me know if that is deemed OK or if you'd require tab completion for this feature as well (currently there is no tab completion for traditional inline splatting, hence why I see it as an additional, optional, nice to have feature). If tab completion is required, I might try to add it but I want to have approval first that the PS wants this for PS7, otherwise I am not going to proceed right now with resolving the merge conflict, etc.

bergmeister committed:
Resolve merge conflicts - Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

# Conflicts:
#	src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs
#	src/System.Management.Automation/engine/parser/token.cs

PoshChan commented on src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs:
```diff
@@ -127,6 +127,9 @@ static ExperimentalFeature()
                 new ExperimentalFeature(
```
@bergmeister, your last commit had 1 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/learning-powershell/powershell-beginners-guide.md.https://examples.oreilly.com/9780596528492/ should work
```powershell
retry of URL failed with error: Connection timed out
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```

bergmeister replied:
Not related to my changes, the URL `https://examples.oreilly.com/9780596528492` is down at the moment.

bergmeister replied:
Fixed now by pulling the fix from master

lzybkr commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -4622,6 +4622,19 @@ internal Token NextToken()
+                            if (c2 == '{')
+                            {
+                                UngetChar();
+                                return NewToken(TokenKind.At);
```
This is worth a short comment that it is putting back `@` (c1). Normally it looks wrong to see UngetChar shortly after PeekChar.

That said, the extra checks here seem unnecessary and getting in the way of fully generalized splatting. At a minimum, the code should have comments rationalizing why - because it seems to me we don't need any additional checks on `c1` here - we just return TokenKind.At straightaway.

bergmeister replied:
I've only added the additional lookahead for `@@{` just to be on the safe side. I've simplified it now as per your request to look ahead only for `@@` and added a comment about why we put the 2nd `@` token back.

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

bergmeister committed:
Reduce lookahead to only @@ to return @ already after having detected @

lzybkr commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -4626,12 +4626,9 @@ internal Token NextToken()
-                                UngetChar();
-                                return NewToken(TokenKind.At);
-                            }
+                            // We have seen 2 @ characters (generalized splatting), therefore putting the 2nd token back
```
I don't think you need to look for `@@` at all - generalized splatting should support `@$var` e.g

bergmeister replied:
Ok, I've removed that lookahead now as well although I feel like by removing the lookahead, it slightly increases the risk that special cases of invalid syntax might behave a bit differently now.
If possible, I'd like to implement only 1 feature of generalized splatting in this PR, which is the syntax `get-foo @@{ Param1 = $Param1Value}`. This is the only feature that I need and is useful to me and it is a standalone feature of the generalized splatting RFC.
The `get-foo @hashtableVariable` syntax already enables using a variable for splatting, therefore I do not see the value in adding the `get-foo @$hashtableVariable` syntax and would rather leave it to someone else if they wanted to in the spirit of open source and keeping it minimum viable.

bergmeister committed:
Do not lookahead for @@. Just return At token at the end.

bergmeister committed:
Enhance semantic check to require more than 1 element in the hashtable to avoid PSReadline throwing an error when only starting to typing '@' in a new line

bergmeister committed:
tweak semantic check

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

KirkMunro commented:
For consideration by anyone reviewing this PR, please have a look at PR #10984 as well, which supports inline splatting as well as other scenarios requested in the original RFC via a `-splat` common parameter.

bergmeister commented:
Thanks, I do not have any objections against your implementation approach.
At the end the committee should make a decision which implementation they favour or if they want a mix of them, I am open to any solution that gets inline splatting into ps 7.

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

# Conflicts:
#	src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs
#	src/System.Management.Automation/engine/parser/token.cs

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into InlineSplatting_AtAtCurly

# Conflicts:
#	src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs

adityapatwardhan commented:
There are two PRs that are implementing the same RFC - [RFC002 ](https://github.com/PowerShell/PowerShell-RFC/blob/master/2-Draft-Accepted/RFC0002-Generalized-Splatting.md). 

I request @PowerShell/powershell-committee to guide on which of the implementations should be accepted.

The other PR with the implementation is: https://github.com/PowerShell/PowerShell/pull/10984

bergmeister commented:
It should be said that both PRs do not implement all the features in the RFC, just a subset of it that me and the other author are the most important for our use cases.
Also: This PR follows the `@@` syntax that was agreed in the RFC exactly, the other PR proposes a different approach via common parameters.

I'd be happy as long as either of the 2 PRs get accepted. When deciding between the 2 PRs, I'd like the @PowerShell/powershell-committee to also consider what is best in terms of follow up PRs if other people contribute e.g. the remaining features in the RFC or tab completion for the inline-splatted arguments.

joeyaiello commented:
We in the @PowerShell/powershell-committee are going to holistically review this vs. #10984 next week in our RFC meeting. This one seems to match the RFC we approved (at least skimming through), while @KirkMunro's does not, but we want to also consider the `-splat` parameter as a possibility. 

That being said, we did review the original RFC0002 and its feedback *very* extensively before accepting.

codykonior commented:
I like -splat, only because -@@{ feels so visually jarring.

vexx32 commented:
Personally, I don't like either very much. `-Splat` feels out of place / could use a better name, perhaps. `@@{ ... }` as you say, feels more like a typo than useful syntax, though its brevity is nice.

I don't have a compelling alternative, but I wish I did.

oising commented:
I prefer `@@{ ... }` -- it's more intuitive to me.

rjmholt commented:
I have to admit that `-@` has the elegance of only applying to commands, so we don't have to think about what `@@{ ... }` means when not being passed to a command as a parameter.

However, I think my personal preference is for a splat sigil or operator:

- Ubiquitous parameters that do special strange things conflate syntax and parameter value passing too much for my liking
- We already have `@var`, which doesn't square well with `-@`; the different sigil implies to me that the shape of the peg has changed, whereas the `-@` implies that the shape of the hole has. My mental schema of splatting is now inconsistent juggling both at once
- We also support multiple splatting today (example below). This makes it harder for `-@`, since currently PowerShell does not allow duplicate parameter names.

```powershell
> $x = 1,2,3
> $y = @{ d = 'a'; e = 'b' }
> $z = @{ f = 'Duck'; g = 'Banana' }
> function Test($a, $b, $c, $d, $e, $f, $g) { $a; $b; $c; $d; $e; $f; $g }
> Test @x @y @z
1
2
3
a
b
Duck
Banana
```

vexx32 commented:
@rjmholt `-@` is actually never parsed as a parameter name currently, so it would need changes to parsing as it is. It parses as a positional value only at the moment. In my opinion `-@` would be nicer but I understand everyone has their own preferences.

rjmholt commented:
> -@ is actually never parsed as a parameter name currently

Yeah I get that, saw above you said it currently parses as a bareword literal. But my concerns are both more and less technical than that:

- Changing parsing would need to be done either in the parser (probably the wrong place) or the parameter binder (more likely), and that would be difficult to implement correctly.
- More importantly, irrespective of how the implementation treats it, the language is a human/computer interaction interface, and the question in my mind is how users who are used to PowerShell will parse it. It looks like any other parameter, just with a strange name. But to what extent will it obey the existing parameter conventions? I think the fact we would need a very different code path to deal with it is a warning sign that it would probably open up a number of cases where what users expect and what it does differ.

bergmeister committed:
Merge branch 'master' of http://github.com/powershell/powershell into InlineSplatting_AtAtCurly

# Conflicts:
#	src/System.Management.Automation/engine/ExperimentalFeature/ExperimentalFeature.cs
#	src/System.Management.Automation/engine/parser/token.cs
#	test/powershell/Language/Parser/Parsing.Tests.ps1

bergmeister committed:
Merge branch 'InlineSplatting_AtAtCurly' of http://github.com/bergmeister/powershell into InlineSplatting_AtAtCurly

bergmeister committed:
Fix merge error in 9b0c1a93be5a7a1ccff4e039137a9165014bc61d

KirkMunro commented:
> We also support multiple splatting today (example below). This makes it harder for -@, since currently PowerShell does not allow duplicate parameter names.

@rjmholt That wouldn't be an issue if -@ accepted an array, like the -splat PR already does.

KirkMunro commented:
Also this discussion goes a little further than simply the syntactical differences between `@@{...}` and `-splat @{}`. The latter also supports splatting properties or method results, such as:

```powershell
function Do-This {
    [CmdletBinding()]
    param(
        [string]$Value
    )
    $sb = {
        # $PSBoundParameters is empty in this child scope, but $PSCmdlet.MyInvocation.BoundParameters is not
        Invoke-Something -splat $PSCmdlet.MyInvocation.BoundParameters
    }
    & $sb
}
```

bergmeister commented:
@PoshChan please retry macos

PoshChan commented:
@bergmeister, did not find matching build context: `PowerShell-CI-macOS`; allowed contexts: PowerShell-CI-SSH

bergmeister committed:
empty commit to rerun ci

bergmeister committed:
empty commit to rerun ci

SteveL-MSFT commented:
@PowerShell/powershell-committee discussed this in context of [RFC0002](https://github.com/PowerShell/PowerShell-RFC/blob/master/2-Draft-Accepted/RFC0002-Generalized-Splatting.md) where we decided that we would withdraw RFC0002 with the exception of fixing the issue to allow explicit parameters to override a splatted hashtable and discuss named parameters for .NET method invocation as a separate issue.  So in light of that, we will be closing this PR as discussed with @bergmeister

SeidChr commented:
no -splat, no -@, no @@ yet, will there be anything like that in the future? It seems like all approaches to improve splatting and thus usability are declined. Is there anything else in the work?
